### ISSUE TITLE
feat: access raw request.body

### ISSUE THREAD

**Issue by bradenmacdonald** [drashland/drash#604]
## Summary

I would like to allow API consumers to upload a file directly in a request body, not via form encoding. In particular, I want to be able to process large file uploads as a stream. It seems that currently Drash always consumes `request.body`, so trying to access it myself results in the error `ReadableStream is locked.`

Why:

Some REST API methods make more sense as direct file uploads in the body, rather than attaching form-encoded files.

Some REST API methods should be able to handle huge (GB+) file uploads as streams, rather than storing file data in memory.

## Example Code

Most likely a flag would have to be set on the `Resource` or the method, to tell Drash not to process the body.

What I want to do:

```
const contentType = request.headers.get("Content-Type");
const bodyStream = request.body;
// Process the stream...
```

What I have to do for now as a workaround:

```
const formFile = request.bodyParam<{content: string, type: string, size: number}>('file');
const contentType = formFile.type;
const bodyStream = readableStreamFromIterable([ new TextEncoder().encode(formFile.content) ]);
// Process the "stream"...
```

**Comment by bradenmacdonald** [comment]
Actually the current situation may be more of an issue than I thought. Today I tried to deal with uploaded image files, but kept finding that their data was corrupted. I believe the current use of `FileReader.readAsText()` corrupts binary data, resulting in a string that cannot be used to recover the originally uploaded binary data. (Because it's not a reversible transformation if the original data contained non-UTF8 bytes).

I had to make this change to Drash: https://github.com/neolace-dev/drash/commit/d387d5f346bf2553ba324640b8f06e93884b9643 in order to be able to handle image uploads. Or am I missing something - is there a way to get the original binary data uploaded by users with the current API?

**Comment by ebebbington** [comment]
Thanks @bradenmacdonald, we’ll try look into it and see if there’s a way you can get the raw body and investigate this issue

**Comment by ebebbington** [comment]
@bradenmacdonald yeah you are right, we weren't reading files properly, and i switched in your code and it all works fine :) If we added this fix, would you still want (have a usecase) to read the raw body? the team were talking about it and just curious if it is needed (high priority) or if not (then there's no rush to get it done)

**Comment by bradenmacdonald** [comment]
@ebebbington I definitely need a way to read binary files like images, so thanks for making that fix!

But otherwise I'd say there's no rush for accessing the raw request.body. Although it's nice to be able to handle large uploads as a stream and to allow API designs that involve POSTing files directly as the body of a request, I wouldn't say it's urgent. At least for my use cases - I'm mostly dealing with image uploads, not videos or anything like that. Thanks for asking and for being so responsive!

**Comment by ebebbington** [comment]
@bradenmacdonald ok perfect, well you're in luck, we were just discussing a plan to move the parse budy logic into a service that the user passes to their server if they want bodies parsed, still a discussion but could be v3 following this fix (v2.5.2)

**Comment by ebebbington** [comment]
Leaving this issue open for support of accessing raw body, but #606 will fix the issue of corrupted files

**Comment by ebebbington** [comment]
@bradenmacdonald 2.5.2 release, feel free to check the latest release

**Comment by bradenmacdonald** [comment]
@ebebbington 2.5.2 works great for my image upload use case. Thanks so much!

**Comment by crookse** [comment]
@bradenmacdonald , so i tried making the request body accessible, but it doesn't seem easily doable lol. every solution i tried affected core, which is something we try not to do. seems like we lost the ability to override drash internal classes in v2. in v1, we were able to do:

```typescript
import { Drash }  from "..."

// Create a new request class that can be used internally in Drash
MyNewRequestClass extends Drash.Request { ... }

// Redefine the internal request class with the new one (Drash will now use this inernally)
Drash.Request = MyNewRequestClass;
```

the above solution would've allowed for you to extend `Drash.Request`, override it's `static async create(...)` method, and then redefine `Drash.Request` with your new request class (maybe like ~10 lines of code). im finding it impossible (at least right now) to be able to do that ha... we might be able to introduce a service that replaces the drash internals to easily do this. need thoughts from the team tho. @ebebbington @Guergeiro, is this possible without affecting core?

**Comment by Guergeiro** [comment]
Honestly, for the sake of not touching the core behaviour, I would just add another prop to the `Request`. Something along the lines of `request.raw`. We just need to make sure we put it in there before parsing it.

Is it the best approach? Nope, as we are polluting the `Request` with yet another property that doesn't have a meaningful relationship to `Request`. Ideally the body would be always raw and if the user wants to have it parsed, we provide a service that can be injected per endpoint, resource or at server level that does the job, or the user can parse it manually how it wants.

Hell, take the [`fetch`](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) api for example. The response is always raw, you just ask it to be parsed.
```ts
const res = await fetch("https://some-url.com"); // Res contains raw body and stuff
const parsedJson = await res.json(); // We want JSON format, so parse it for us
```

In sum:

- `request.rawBody` - solves the issue without changing core, keeping it v2.
- `request.body` - always returns raw and the user opts-in to the parser, for future v3.

**Comment by crookse** [comment]
very true. i didn't think about response.json() or .text() in the fetch API. maybe that's something we move towards in the next major version cause that really does make sense. for now i do agree on another prop

**Comment by bradenmacdonald** [comment]
Hey folks, today I tried updating my app to v2.7.0 and using the new `request.original` to be able to handle uploaded files in the POST body (as streams, because I want streaming upload of large files). I found that the `request.original` is working great - *if* the `Content-Type` being uploaded is `text/plain`, `application/json`, or one of the types that Drash knows how to parse. But if I try to let my API accept other content types in the body, e.g. uploading an `image/png` as the request body, [Drash rejects it with an HTTP 402 error](https://github.com/drashland/drash/blob/30d98f036efee30b8326d702f0c02f28c2bb7624/src/http/request.ts#L222-L242).

Should I open a new feature request, something like "tell Drash not to parse the body for certain requests"?
